### ISSUE TITLE
sails-mysql: populate adding order by when sort is not set and breaking the query

### ISSUE THREAD

**Issue by sgress454** [balderdashy/sails#3992]
_From @mneil on February 1, 2017 18:55_

I created a has many relationship. Example models are incomplete but show the relationship definition

passport.js
```
module.exports = { attributes: {
 user: { model: 'user', required: true },
} };
```

user.js
```
module.exports = { attributes: {
 passports: {
      collection: 'passport',
      via: 'user'
    },
} };
````

Doing a populate on users of passports returns empty passports array
```
sails.models.user.findOne()
      .where(_user)
      .populate('passports').then(user => console.log(user));
```

The relationship exists. So I dug through the code and found out I could enable query logging. Here are the queries sails-mysql are running:
```
MySQL.populateBuffers:  SELECT `user`.`created_at`, `user`.`updated_at`, `user`.`id`, `user`.`name`, `user`.`username`, `user`.`email` FROM `user` AS `user`  WHERE `user`.`email` = "test@test.com"
MySQL.processChildren:  (SELECT * FROM `passport` AS `passport` WHERE `user` = 5  ORDER BY `passport`.`undefined` ASC)
```

Where did the order by in the passport query come from? Explicitly setting a sort in the query makes it return the correct data.

```
sails.models.user.findOne()
      .where(_user)
      .populate('passports',{sort: 'updated_at DESC'})
```

sails: 0.12.11
sails-mysql: 0.11.5

_Copied from original issue: balderdashy/sails-mysql#341_

**Comment by sgress454** [comment]
_From @sailsbot on February 1, 2017 18:55_

@mneil Thanks for posting, we'll take a look as soon as possible.  In the meantime, if you haven’t already, please carefully read the [issue contribution guidelines](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md#ii-issue-contributions) and double-check for any missing information above.  In particular, please ensure that this issue is about a stability or performance bug with a  documented feature; and make sure you’ve included detailed instructions on how to reproduce the bug from a clean install.   Finally, don’t forget to include the version of Node.js you tested with, as well as your version of Sails or Waterline, and of any relevant standalone adapters/generators/hooks.

Thank you!

**Comment by sgress454** [comment]
_From @mneil on February 1, 2017 22:38_

More info on the problem:
I have automatic id generation set to false. I do not have an id set on my passport.js model. Apparently the adapter always tries to sort by id (which is undefined) and that causes the error. 

I do not see the need for the automatic sort though.

**Comment by sgress454** [comment]
_From @particlebanana on February 2, 2017 21:46_

@mneil what's the primary key on the `passport` model? You don't need to have an `id` attribute set but each model needs a unique primary key.

The sorting is so that you don't get back random result sets each query. All queries do a default sort using the primary key if a sort is not defined.

**Comment by sgress454** [comment]
_From @mneil on February 3, 2017 2:28_

I am using MySQL and have an id in the database set to the primary key w/ auto increment. But, I did not have a primary key defined in sails. The passport model belongs to the user on user id, there can be only 1 unique combination of passport w/ user id and type, and the table cascade deletes on user removal.

This is probably an edge case but it's worth mentioning that a primary key must be defined for populate to work. I'm also still confused by why the sort is there at all.

**Comment by sgress454** [comment]
@mneil There doesn't need to be a primary key defined in Sails (at least not in Sails v0.12.11) unless `autoPK` is set to `false` for a model.  If what you posted for your model files is complete, then Sails should be creating `id` columns for the `user` and `passport` tables for you.  As for the sort, as @particlebanana said, it's there because mysql doesn't guarantee row order if you don't specify a sort, so in Waterline we enforce that they always come back ordered by primary key so that you get consistent results.  Why you're seeing 
```
`ORDER BY `passport`.`undefined`
```
is a mystery, as well as why the process isn't just crashing when trying to run that query, since I'm assuming you don't have a column named `undefined` in your table.  

I'm not able to reproduce this with a new app, but if you're still seeing it, could you create a new repo that reproduces the problem for us to look at?

e.g.:

```
sails new sails-mysql-issue-341
cd sails-issue-341
npm install sails-mysql@0.11.5 --save
# Now make whichever are the simplest/quickest changes which reproduce the issue,
# and when you for sure have the issue isolated, run:
git init
hub create  # or use the GitHub UI to create the repo and `git remote add` it manually
git add .
git commit -am 'Reproduces https://github.com/balderdashy/sails/issues/3499'
git push -u origin master
hub browse # or open the repo in your browser manually
# Then copy the URL of the new repo and paste it as a comment on this issue.
# We'll check it out, and since this way we'll be able to experience the problem
# ourselves, we should have no problem getting to the bottom of it.
```

Thanks!

p.s. we're consolidating all of the issues under Sails, so this issue may be closed and moved over momentarily.  Just post any new information in the new issue.  Thanks again!

**Comment by mneil** [comment]
There is no primary key. Auto primary key is false. I have not defined a primary key in sails. Passports are searched by user and type and neither are unique.

Primary keys are not required by the engine nor are they needed for every schema design. Indexes, yes, but I do not have or need a primary key on this table.

Again, I'm likely an edge case here. But, my question still stands as to why sort would exist.

Many people use guid for mongo (a supported adapter) and likely as a primary key. These are also not sortable and do not guarantee order. I argue that the user should be left to sort on their own and not have the ORM add in additional parameters like this.

**Comment by sgress454** [comment]
Ah, understood.  It's definitely an edge case, and while it's certainly valid to not have a primary key in your database table, Sails just doesn't operate this way.  Requiring your models to have a primary key (even if you don't use it) is a small tradeoff for the benefits that the ORM provides.

The real mystery to me is how 
```
ORDER BY `passport`.`undefined`
```

isn't just throwing an error.  But in Sails 1.0 this is moot, since your app won't lift without a primary key being defined for each model.

> Many people use guid for mongo (a supported adapter) and likely as a primary key. These are also not sortable and do not guarantee order.

If you're referring to ObjectIDs, they are in fact sortable, and doing so gives you the records back in insertion order.  In the general case though, Sails adds a sort clause to the population query not guarantee that the records come back in a particular order, but to guarantee that two identical calls to `.find()` will return records in the same order both times.  You can always override the particular sort clause used by providing one yourself in the call to `.populate()`.

**Comment by mneil** [comment]
I am not referring to ObjectIDs, I'm referring to GUID or UUID if you prefer. Those are not sortable.

Thank you for opening this ticket again. Even if you do not agree with me it would be at least nice to not silently order by undefined and not return records.

**Comment by sailsbot** [comment]
@sgress454,@mneil: Hello, I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message and simply close the issue if possible.  On the other hand, if you are still waiting on a patch, please post a comment to keep the thread alive (with any new information you can provide).



If no further activity occurs on this thread within the next 3 days, the issue will automatically be closed.

Thanks so much for your help!
